File: 5/5_11.py
def str2dict(s: str) -> dict[str, int]:
    if s.startswith('{') and s.endswith('}'):
        s = s[1:-1]
    if len(s) == 0:
        return dict({})
    l = s.split(',')
    dl = []
    for i in l:
        i.replace(' ', '')
        dl.append((i.split(':')[0], int(i.split(':')[1])))
    return dict(dl)
def solve_problem(d1: dict[str, int], d2: dict[str, int]) -> str:
    d1keys = list(d1.keys())
    d2keys = list(d2.keys())
    adds = []
    dels = []
    mods = []
    for i in d2keys:
        if i not in d1keys:
            adds.append(i)
    for i in d1keys:
        if i not in d2keys:
            dels.append(i)
    for i in d1keys:
        if i in d2keys:
            if d1[i] != d2[i]:
                mods.append(i)
    s = ""
    adds.sort()
    dels.sort()
    mods.sort()
    if len(adds) != 0:
        s += "+"
        s += ",".join(adds)
        s += ";"
    if len(dels) != 0:
        s += "-"
        s += ",".join(dels)
        s += ";"
    if len(mods) != 0:
        s += "*"
        s += ",".join(mods)
        s += ";"
    if len(s) == 0:
        s = "No changes"
    return s
def Solve():
    s1 = input()
    s2 = input()
    print(solve_problem(str2dict(s1), str2dict(s2)))
Solve()

File: 5/5_11.cpp
#include <iostream>
#include <unordered_map>
#include <vector>
#include <sstream>
#include <algorithm>

using namespace std;

// Function to convert a string of form "{a:1, b:2}" to a dictionary (unordered_map)
unordered_map<string, int> str2dict(string s) {
    unordered_map<string, int> dict;
    // Remove the starting and ending curly braces
    if (s.front() == '{' && s.back() == '}') {
        s = s.substr(1, s.size() - 2);
    }
    if (s.empty()) {
        return dict;
    }
    
    stringstream ss(s);
    string token;

    while (getline(ss, token, ',')) {
        token.erase(remove(token.begin(), token.end(), ' '), token.end()); // Remove spaces
        size_t colon_pos = token.find(':');
        string key = token.substr(0, colon_pos);
        int value = stoi(token.substr(colon_pos + 1));
        dict[key] = value;
    }

    return dict;
}

// Function to compare two dictionaries and return the changes
string solve_problem(unordered_map<string, int>& d1, unordered_map<string, int>& d2) {
    vector<string> adds, dels, mods;
    vector<string> d1keys, d2keys;

    for (const auto& pair : d1) {
        d1keys.push_back(pair.first);
    }
    for (const auto& pair : d2) {
        d2keys.push_back(pair.first);
    }

    // Check for added keys
    for (const auto& key : d2keys) {
        if (d1.find(key) == d1.end()) {
            adds.push_back(key);
        }
    }

    // Check for deleted keys
    for (const auto& key : d1keys) {
        if (d2.find(key) == d2.end()) {
            dels.push_back(key);
        }
    }

    // Check for modified keys
    for (const auto& key : d1keys) {
        if (d2.find(key) != d2.end() && d1[key] != d2[key]) {
            mods.push_back(key);
        }
    }

    // Sort all lists for consistent output
    sort(adds.begin(), adds.end());
    sort(dels.begin(), dels.end());
    sort(mods.begin(), mods.end());

    string result = "";

    // Format the result string
    if (!adds.empty()) {
        result += "+";
        for (size_t i = 0; i < adds.size(); ++i) {
            result += adds[i];
            if (i != adds.size() - 1) result += ",";
        }
        result += ";";
    }
    if (!dels.empty()) {
        result += "-";
        for (size_t i = 0; i < dels.size(); ++i) {
            result += dels[i];
            if (i != dels.size() - 1) result += ",";
        }
        result += ";";
    }
    if (!mods.empty()) {
        result += "*";
        for (size_t i = 0; i < mods.size(); ++i) {
            result += mods[i];
            if (i != mods.size() - 1) result += ",";
        }
        result += ";";
    }

    if (result.empty()) {
        result = "No changes";
    }

    return result;
}

int main() {
    string s1, s2;

    // Input two dictionary-like strings
    getline(cin, s1);
    getline(cin, s2);

    // Convert the strings to dictionaries
    unordered_map<string, int> d1 = str2dict(s1);
    unordered_map<string, int> d2 = str2dict(s2);

    // Print the result of the comparison
    cout << solve_problem(d1, d2) << endl;

    return 0;
}
